charms/knative-eventing/src/charm.py

```diff
@@ -71,6 +71,27 @@ def _on_install(self, _):
     def _on_config_changed(self, _):
         self._apply_and_set_status()
 
+    def _get_relation_data(self, relation_name: str, source):
+        """Returns relation data from a specified relation name and source.
+
+        Args:
+            relation_name: the name of the endpoint for this charm
+            source: the application or unit whose data will be retrieved
+        """
+        relation = self.model.get_relation(relation_name)
+        return relation.data[source]
+
+    def _on_otel_collector_relation_changed(self, event):
+        """Event handler for on['otel-collector'].relation_changed."""
+        # Read relation data from knative-operator application bucket
+        otel_collector_rel_data = self._get_relation_data("otel-collector", event.app)
+        if otel_collector_rel_data:
+            # Extend context to render manifests
+            self._context_ext = otel_collector_rel_data
+            # Reset resource handler
+            self._resource_handler = None
+            self._apply_and_set_status()
```

Comment:
I think this needs to be moved to `_apply_and_set_status()`.  Consider:

* install hook fires
  * KRH is rendered without otel data, applied, and now CR exists without obs data (as expected)
* we relate to operator and get the otel data back
  * KRH is rendered with the otel data, applied, and now CR has obs data (as expected)
* config is changed
  * KRH is rendered *without the relation data*, applied, and now CR no longer has obs data (as expected)

If the relation data exists, we **always** want to use it, even if we're updating our resources based on some other event.  by moving this into `_apply_and_set_status` (or really, by moving this logic into a helper `_get_otel_data` and calling the helper in `_apply_and_set_status`) we always get the fully rendered configuration.  Then we just either handle the relation_changed event with the same handler as config or install, or have a little stub handler like those ones that calls `_apply_and_set_status`.

Reply:
Or, just put the `_get_otel_data` call into `self.context`.  I don't think anything else needs the data, so just putting the call in `self.context` means nobody else even knows there's a relation in the first place.  They just know they ask for context, and context figures it out for them.

Reply:
I have updated the way we populate `self.context`. Please let me know if this suffices.